include/JSystem/JStudio/JStudio/functionvalue.h

```diff
@@ -1,6 +1,392 @@
+
+private:
+    /* 0x00 */ TFunctionValueAttribute_refer* refer_;
+    /* 0x04 */ TFunctionValueAttribute_range* range_;
+    /* 0x08 */ TFunctionValueAttribute_interpolate* interp_;
+};
+
+class TFunctionValueAttributeSet : public TFunctionValueAttributeSet_const {
+public:
+    TFunctionValueAttributeSet(TFunctionValueAttribute_refer* refer,
+                               TFunctionValueAttribute_range* range,
+                               TFunctionValueAttribute_interpolate* interp)
+        : TFunctionValueAttributeSet_const(refer, range, interp) {}
+
+    TFunctionValueAttribute_refer* refer_get() const {
+        return static_cast<const TFunctionValueAttributeSet_const*>(this)->refer_get();
+    }
+    TFunctionValueAttribute_range* range_get() const {
+        return static_cast<const TFunctionValueAttributeSet_const*>(this)->range_get();
+    }
+    TFunctionValueAttribute_interpolate* interpolate_get() const {
+        return static_cast<const TFunctionValueAttributeSet_const*>(this)->interpolate_get();
+    }
+};
+
+class TFunctionValueAttribute_refer : public JGadget::TVector_pointer<TFunctionValue*> {
+public:
+    // TFunctionValueAttribute_refer() :
+    // JGadget::TVector_pointer<TFunctionValue*>(JGadget::TAllocator<void*>()) {}
+    ~TFunctionValueAttribute_refer() {}
+
+    /* 802816E8 */ void refer_initialize();
+
+    const TFunctionValueAttribute_refer* refer_getContainer() const { return this; }
+    JGadget::TVector_pointer<TFunctionValue*>& refer_referContainer() { return *this; }
+    bool refer_isReferring(const TFunctionValue* p) const { return false; }  // todo
+};
+
+class TFunctionValueAttribute_range {
+public:
+    /* 80281D18 */ TFunctionValueAttribute_range();
+
+    /* 80281918 */ void range_initialize();
+    /* 8028194C */ void range_prepare();
+    /* 802819F4 */ void range_set(f64, f64);
+    /* 80281A08 */ f64 range_getParameter(f64, f64, f64) const;
+
+    TFunctionValue::TEProgress range_getProgress() const {
+        return (TFunctionValue::TEProgress)mProgress;
+    }
+    void range_setProgress(TFunctionValue::TEProgress progress) { mProgress = progress; }
+    TFunctionValue::TEAdjust range_getAdjust() const { return (TFunctionValue::TEAdjust)mAdjust; }
+    void range_setAdjust(TFunctionValue::TEAdjust adjust) { mAdjust = adjust; }
+    void range_setOutside(TFunctionValue::TEOutside outside) { range_setOutside(outside, outside); }
+    void range_setOutside(TFunctionValue::TEOutside begin, TFunctionValue::TEOutside end) {
+        range_setOutside_begin(begin);
+        range_setOutside_end(end);
+    }
+    void range_setOutside_begin(TFunctionValue::TEOutside begin) { mBegin = begin; }
+    void range_setOutside_end(TFunctionValue::TEOutside end) { mEnd = end; }
+    f64 range_getParameter_outside(f64 arg1) const {
+        f64 temp = arg1 - fBegin_;
+        f64 result = temp;
+        if (temp < *(f64*)&lit_652) {
```

Comment:
```suggestion
        if (temp < DOUBLE_LABEL(/* 0.0 */ lit_652)) {
```
I usually put the value of the label there so it's easier to see what it's doing